stonesoup/metricgenerator/uncertaintymetric.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+from .base import MetricGenerator
+from ..types.state import State, StateMutableSequence
+from ..types.metric import SingleTimeMetric, TimeRangeMetric
+from ..types.time import TimeRange
+
+
+class UncertaintyMetric(MetricGenerator):
+    """
+    Computes the covariance matrix of the state for each track as a measure of the
+    uncertainty of the track. The metric is calculated as the sum of all covariance
+    matrix norms at each time step.
+    """
+
+    def compute_metric(self, manager):
+        """Computes the metric using the data in the metric manager
+
+        Parameters
+        ----------
+        manager : :class:`~.MetricManager`
+            Contains the data to be used to create the metric
+
+        Returns
+        -------
+        metric : list :class:`~.Metric`
+            Containing the metric information. The value of the metric is a
+            list of the metric at each timestamp
+
+        """
+
+        return self.compute_over_time(self.extract_states(manager.tracks))
+
+    @staticmethod
+    def extract_states(object_with_states):
+        """
+        Extracts a list of :class:`~states` from a list of (or single) objects
+        containing states. This method is defined to handle :class:`~track`,
+        :class:`~groundtruthpath` and :class:`~detection` objects
```

Comment:
Rather this is defined to handle sets of `StateMutableSequence` and `State` types.